### PR TITLE
Upgrade to MOI 1.21.0 and switch to SolutionLimit

### LINKED ISSUE #45: Use new MOI SolutionLimit attribute
instead of `MOI.RawOptimizerAttribute("num_solutions")`, after MOI is tagged. See https://github.com/jump-dev/MathOptInterface.jl/pull/2291

### PR DESCRIPTION
**Pull request by zengjian-hu-rai**
Replace all "num_solutions" by the new MOI.SolutionLimit

Closes #45

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -1040,6 +1040,12 @@ function test_moi_basic_fzn()
     return
 end
 
+function test_moi_support_solution_limit()
+    solver = MiniZinc.Optimizer{Int}("chuffed")
+    MOI.supports(solver, MOI.SolutionLimit())
+    @test MOI.get(solver, MOI.SolutionLimit()) == 1
```
Comment (odow) [inline test/runtests.jl:1046]:
```suggestion
    @test MOI.get(solver, MOI.SolutionLimit()) == 1
    return
```

File: src/optimize.jl [hunk 2]
```diff
@@ -131,10 +144,6 @@ function MOI.get(model::Optimizer, attr::MOI.RawOptimizerAttribute)
 end
 
 function MOI.set(model::Optimizer, attr::MOI.RawOptimizerAttribute, value)
-    if attr.name == "num_solutions" && !(value isa Int && value >= 1)
-        msg = "value must be an `Int` that is >= 1"
-        throw(MOI.SetAttributeNotAllowed(attr, msg))
-    end
```
Comment (odow) [inline src/optimize.jl:137]: I think it's best to keep this as-is. Removing the option is breaking

File: src/optimize.jl [hunk 3]
```diff
@@ -231,7 +240,7 @@ function MOI.get(model::Optimizer, ::MOI.TerminationStatus)
     if model.solver_status == "UNSATISFIABLE"
         return MOI.INFEASIBLE
     elseif _has_solution(model)
-        if 1 < model.options["num_solutions"] <= length(model.primal_solutions)
+        if 1 < model.options["solution_limit"] <= length(model.primal_solutions)
```
Comment (odow) [inline src/optimize.jl:243]: Let's keep this as `num_solutions`

File: src/optimize.jl [hunk 4]
```diff
@@ -122,6 +124,17 @@ function MOI.empty!(model::Optimizer{T}) where {T}
...
 
+MOI.supports(::Optimizer, ::MOI.SolutionLimit) = true
+MOI.get(model::Optimizer, ::MOI.SolutionLimit) = model.options["solution_limit"]
+
+function MOI.set(model::Optimizer, attr::MOI.SolutionLimit, value)
+    if !(value isa Int && value >= 1)
+        msg = "[MiniZinc] SolutionLimit must be an `Int` that is >= 1"
+        throw(MOI.SetAttributeNotAllowed(attr, msg))
+    end
+    return model.options["solution_limit"] = value
```
Comment (odow) [inline src/optimize.jl:135]:
```suggestion
    model.options["solution_limit"] = value
    return
```

File: src/optimize.jl [hunk 5]
```diff
@@ -83,9 +84,10 @@ function _run_minizinc(dest::Optimizer)
             limit = round(Int, 1_000 * dest.time_limit_sec::Float64)
             cmd = `$cmd --time-limit $limit`
         end
-        if dest.options["num_solutions"] > 1
-            cmd = `$cmd --num-solutions $(dest.options["num_solutions"])`
+        if dest.options["solution_limit"] > 1
+            cmd = `$cmd --num-solutions $(dest.options["solution_limit"])`
         end
+        @info "[MiniZinc] Cmd = $cmd"
```
Comment (odow) [inline src/optimize.jl:90]:
```suggestion
```

**Comment by odow** [comment]
We also need to change the compat bounds in `Project.toml`

File: src/optimize.jl (hunk 2 above)
Comment (zengjian-hu-rai) [reply src/optimize.jl]: sure, make sense

**Review by odow — APPROVED** [review]
I simplified things even further

**Pull request merged**